### PR TITLE
Address missing functionality in call opcode

### PR DESCRIPTION
**Pull request by Voxelot**
There were several blocking todo's preventing contract calls from working properly, this PR should address all of them.

### CODE REVIEW

File: src/interpreter/flow.rs [hunk 1]
```diff
@@ -104,9 +104,7 @@ where
 
         self.memory[self.registers[REG_FP] as usize..self.registers[REG_SP] as usize].copy_from_slice(stack.as_slice());
 
-        // TODO set balance for forward coins to $bal
-        // TODO set forward gas to $cgas
-
+        self.registers[REG_BAL] = b;
```
Comment (vlopes11) [inline src/interpreter/flow.rs:107]: According to the specs, it should be `d`. Maybe typo there? https://github.com/FuelLabs/fuel-specs/blob/a65e14dfe5c0049ebf4f8e0b33b061c60927d7d9/specs/vm/opcodes.md?plain=1#L1098

File: src/interpreter/flow.rs [hunk 2]
```diff
@@ -86,8 +86,8 @@ where
             return Err(PanicReason::ContractNotInInputs.into());
         }
 
-        // TODO validate external and internal context
-        // TODO update asset ID balance
+        // credit contract asset_id balance
```
Comment (vlopes11) [inline src/interpreter/flow.rs:89]: Need to add the internal context part:
https://github.com/FuelLabs/fuel-specs/blob/a65e14dfe5c0049ebf4f8e0b33b061c60927d7d9/specs/vm/opcodes.md?plain=1#L1091

**Review by vlopes11 — CHANGES_REQUESTED** [review]

File: src/interpreter/flow.rs (hunk 1 above)
Comment (vlopes11) [reply src/interpreter/flow.rs:107]: cc @adlerjohn

Comment (adlerjohn) [reply src/interpreter/flow.rs:107]: Good point, looks like a typo in the specs!

Comment (adlerjohn) [reply src/interpreter/flow.rs:107]: https://github.com/FuelLabs/fuel-specs/pull/292

File: src/interpreter/flow.rs (hunk 2 above)
Comment (Voxelot) [reply src/interpreter/flow.rs:92]: done and unit tested

**Comment by Voxelot** [comment]
ah thought it was ready but there's one test regression, fix incoming

File: src/interpreter/flow.rs [hunk 3]
```diff
@@ -80,14 +80,17 @@ where
 
         if self.is_external_context() {
             self.external_asset_id_balance_sub(&asset_id, b)?;
+        } else {
```
Comment (adlerjohn) [inline src/interpreter/flow.rs:83]: Nit, but would this be a good candidate for a `match` instead of `if`/`else`?

**Review by adlerjohn — APPROVED** [review]

File: tests/contract.rs [hunk 4]
```diff
@@ -230,6 +230,169 @@ fn mint_burn() {
...
+
+    // Ensure transfer tx processed correctly
+    assert!(!transfer_tx.should_revert());
+
+    // verify balance transfer occurred
+    let end_balance = test_context.get_contract_balance(&contract_id, &asset_id);
+    assert_eq!(end_balance, call_amount);
+
+    // verify balance register was set
+    assert_eq!(transfer_tx.receipts()[1].val().unwrap(), call_amount);
```
Comment (vlopes11) [inline tests/contract.rs:296]: The logic seems correct, but its always safer to just `LOG` the register and then inspect the receipt. The `CALL` receipt is built with the arguments of the call and by itself isn't a guarantee that the register was properly set.

**Review by vlopes11 — APPROVED** [review]
LGTM

File: src/interpreter/flow.rs (hunk 3 above)
Comment (Voxelot) [reply src/interpreter/flow.rs:83]: Potentially, although the check for is_external is pretty lightweight itself (`if self.registers[REG_FP] == 0`)

**Pull request merged**